File: examples/example_program_stdin.rs
use std::env;
use std::io::{self, BufRead};

fn main() {
    let mut is_uppercase = false;
    let mut intentional_crash = false;
    let mut use_lists = false;
    let mut interactive = false;
    for arg in env::args().skip(1) {
        if arg == "--uppercase" {
            is_uppercase = true;
        } else if arg == "--interactive" {
            interactive = true;
        } else if arg == "--intentional-crash" {
            intentional_crash = true;
        } else if arg == "--use-lists" {
            use_lists = true;
        }
    }

    if interactive {
        let stdin = io::stdin();
        let mut buf = Vec::new();

        println!("start");

        for line in stdin.lock().lines() {
            let Ok(line) = line else { break };

            if line == "close" {
                if !buf.is_empty() {
                    eprintln!("no end to message {buf:?}");
                }
                break;
            }

            if line == "end" {
                let output = String::from_utf8_lossy(&buf);
                {
                    for line in output.lines() {
                        if intentional_crash && line.trim_end().ends_with("2") {
                            panic!("CRASH!!!");
                        }
                        let output = if is_uppercase {
                            std::borrow::Cow::Owned(line.to_uppercase())
                        } else {
                            std::borrow::Cow::Borrowed(line)
                        };
                        if use_lists {
                            print!("- ");
                        }
                        if line.trim_end().ends_with("on stderr") {
                            std::thread::sleep(std::time::Duration::from_millis(10));
                            eprintln!("{output}");
                            std::thread::sleep(std::time::Duration::from_millis(10));
                        } else {
                            println!("{output}");
                        }
                    }
                }
                println!("end");
                buf.clear();
                continue;
            }

            buf.extend_from_slice(line.as_bytes());
            buf.push(b'\n');
        }
    } else {
        let value = env::args().nth(1).unwrap();
        for line in value.lines() {
            let output = if is_uppercase {
                std::borrow::Cow::Owned(line.to_uppercase())
            } else {
                std::borrow::Cow::Borrowed(line)
            };
            if use_lists {
                print!("- ");
            }
            println!("{output}");
        }
    }
}

File: examples/example_program_stdin.js
let uppercase = false, intentionalTimeout = false, intentionalCrash = false, interactive = false, useLists = false;

for (const arg of process.argv.slice(2)) {
	if (arg === "--uppercase") uppercase = true;
	else if (arg === "--intentional-timeout") intentionalTimeout = true;
	else if (arg === "--intentional-crash") intentionalCrash = true;
	else if (arg === "--interactive") interactive = true;
	else if (arg === "--use-lists") useLists = true;
}

// console.error({ uppercase, intentionalTimeout, intentionalCrash, useLists });

if (!interactive) {
	let value = process.argv[2];
	if (useLists) {
		value = value.split("\n").map(line => `- ${line}`).join("\n");
	}
	console.log(uppercase ? value.toUpperCase() : uppercase);
} else {
	const wait = (duration = 1000) => new Promise((res, _rej) => setTimeout(res, duration));

	async function sendMessage(total) {
		if (total.trimEnd().endsWith("2")) {
			if (intentionalTimeout) await wait(3000);
			if (intentionalCrash) throw Error("CRASH!!!");
		}

		const bothChannels = total.includes("stderr");

		for (const line of total.split("\n")) {
			const chunk = uppercase ? line.toUpperCase() : line;
			const stream = line.endsWith("on stderr") ? Bun.stderr : Bun.stdout;

			const writer = stream.writer();
			if (useLists) writer.write("- ");
			writer.write(chunk);
			writer.write("\n");
			writer.flush();

			if (bothChannels) await wait(50);
		}
	}

	console.log("start");
	let buffer = "";
	for await (const line of console) {
		if (line == "close") break;

		if (line == "end") {
			await sendMessage(buffer);
			console.log("end");
			buffer = "";
			continue
		}

		buffer += line;
		buffer += "\n";
	}

	console.log("finished");
}
